packages/persistent-merkle-tree/src/hashComputation.ts

```diff
@@ -0,0 +1,216 @@
+    this.pointer = null;
+  }
+
+  get length(): number {
+    return this._length;
+  }
+
+  get totalLength(): number {
+    return this._totalLength;
+  }
+
+  /**
+   * run before every run
+   */
+  reset(): void {
+    // keep this.head object, only release the data
+    this.head.src0 = null as unknown as Node;
+    this.head.src1 = null as unknown as Node;
+    this.head.dest = null as unknown as Node;
+    this.tail = null;
+    this._length = 0;
+    // totalLength is not reset
+    this.pointer = null;
+  }
+
+  /**
+   * Append a new HashComputation to tail.
+   * This will overwrite the existing HashComputation if it is not null, or grow the list if needed.
+   */
+  push(src0: Node, src1: Node, dest: Node): void {
+    if (this.tail !== null) {
+      let newTail = this.tail.next;
+      if (newTail !== null) {
+        newTail.src0 = src0;
+        newTail.src1 = src1;
+        newTail.dest = dest;
+      } else {
+        // grow the list
+        newTail = {src0, src1, dest, next: null};
+        this.tail.next = newTail;
+        this._totalLength++;
+      }
+      this.tail = newTail;
+      this._length++;
+      return;
+    }
+
+    // first item
+    this.head.src0 = src0;
+    this.head.src1 = src1;
+    this.head.dest = dest;
+    this.tail = this.head;
+    this._length = 1;
+    if (this._totalLength === 0) {
+      this._totalLength = 1;
+    }
+    // else _totalLength > 0, do not set
+  }
+
+  /**
+   * run after every run
+   * hashComps may still refer to the old Nodes, we should release them to avoid memory leak.
+   */
+  clean(): void {
+    let hc = this.tail?.next ?? null;
+    while (hc !== null) {
+      if (hc.src0 === null) {
+        // we may have already cleaned it in the previous run, return early
+        break;
+      }
+      hc.src0 = null as unknown as Node;
+      hc.src1 = null as unknown as Node;
+      hc.dest = null as unknown as Node;
+      hc = hc.next;
+    }
+  }
+
+  /**
+   * Implement Iterator for this class
+   */
+  next(): IteratorResult<HashComputation> {
+    if (!this.pointer || this.tail === null) {
+      return {done: true, value: undefined};
+    }
+
+    // never yield value beyond the tail
+    const value = this.pointer;
+    const isNull = value.src0 === null;
+    this.pointer = this.pointer.next;
+
+    return isNull ? {done: true, value: undefined} : {done: false, value};
+  }
+
+  /**
+   * This is convenient method to consume HashComputationLevel with for-of loop
+   * See "next" method above for the actual implementation
+   */
+  [Symbol.iterator](): IterableIterator<HashComputation> {
+    this.pointer = this.head;
+    return this;
+  }
+
+  /**
+   * Not great due to memory allocation.
+   * Mainly used for testing.
+   */
+  toArray(): HashComputation[] {
+    const hashComps: HashComputation[] = [];
+    for (const hc of this) {
+      hashComps.push(hc);
+    }
+    return hashComps;
+  }
+
+  /**
+   * For testing.
+   */
+  dump(): HashComputation[] {
```

Comment:
Is this also needed? seems that either `toArray` or `dump` should exist

Reply:
`toArray()` will return all HashComputation with data while `dump()` returns that + those with no data (some `HashComputation` actually only a place holder for the next item to be added)

I updated the comment for those methods